arch: arm64: rockchip: use lima driver for gpu of rk3528

CodeChenL opened the pull request.

CodeChenL committed:
arch: arm64: rockchip: use lima driver for gpu of rk3528

Signed-off-by: ChenJiali <chenjiali@radxa.com>

RadxaYuntian commented on arch/arm64/boot/dts/rockchip/rk3528.dtsi:
```diff
@@ -1013,6 +1013,13 @@
+			     <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI 91 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI 88 IRQ_TYPE_LEVEL_HIGH>,
 			     <GIC_SPI 89 IRQ_TYPE_LEVEL_HIGH>,
```
这里不还是有88/89么，重复定义

CodeChenL replied:
这是为了libmali和lima通用一个node节点而重复的，他们使用相似而不相同的`interrupt-names`

RadxaYuntian reviewed (changes requested):
另外现在的修改能否和mali保持兼容？还是说也要类似rk3588一样单独备份一个设备树节点？

CodeChenL commented:
> 另外现在的修改能否和mali保持兼容？还是说也要类似rk3588一样单独备份一个设备树节点？

现在的设备树能保持兼容，但是内核默认关闭mali400驱动的config

RadxaYuntian commented:
是说默认关闭了lima么？可能是bsp方面的修改吧。

CodeChenL commented:
> 是说默认关闭了lima么？可能是bsp方面的修改吧。

lima和mali400都关了，也就是开闭源驱动都关了，bsp还需要一笔提交

RadxaStephen commented:
Hi

```
		interrupts = <GIC_SPI 88 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 89 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 86 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 92 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 93 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 91 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 88 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 89 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 86 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 92 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 93 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>,
			     <GIC_SPI 91 IRQ_TYPE_LEVEL_HIGH>;
		interrupt-names = "Mali_GP_IRQ",
				  "Mali_GP_MMU_IRQ",
				  "IRQPP",
				  "Mali_PP0_IRQ",
				  "Mali_PP0_MMU_IRQ",
				  "Mali_PP1_IRQ",
				  "Mali_PP1_MMU_IRQ",
				  "gp",
				  "gpmmu",
				  "pp",
				  "pp0",
				  "ppmmu0",
				  "pp1",
				  "ppmmu1";
```

这里写了重复的中断。

建议再增加一个 gpu 节点。

CodeChenL commented:
@RadxaStephen request a review

RadxaYuntian commented:
另外我发现如果之前没有review的话，可以把人移除reviewer再添加进来

RadxaYuntian commented on drivers/gpu/arm/mali400/mali/linux/mali_memory_dma_buf.h:
```diff
@@ -50,4 +50,6 @@ void mali_dma_buf_unmap_job(struct mali_pp_job *job);
 }
 #endif
 
+MODULE_IMPORT_NS(DMA_BUF);
```
这个应该在mali模块的入口c里面添加，不应该在头文件里面

CodeChenL replied:
在这个文件 https://github.com/radxa/kernel/blob/linux-6.1-stan-rkr4.1/drivers/gpu/arm/mali400/mali/linux/mali_memory_dma_buf.c 加怎么样，主要就是在这封装的dma_buf相关函数

CodeChenL replied:
我看其他gpu驱动也不是在模块入口引入的，也是在类似封装函数文件里引入的

RadxaYuntian replied:
Nvidia的话是在模块入口这边统一定义的，反正不是在头文件里面：https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/nvidia/nv.c#L120-L134

RadxaYuntian commented on drivers/gpu/arm/mali400/mali/linux/mali_osk_mali.c:
```diff
@@ -137,9 +137,9 @@ _mali_osk_errcode_t _mali_osk_resource_initialize(void)
-		res = platform_get_resource_byname(mali_platform_device, IORESOURCE_IRQ, mali_osk_resource_bank[i].irq_name);
-		if (res) {
-			mali_osk_resource_bank[i].irq = res->start;
+		int irq = platform_get_irq_byname_optional(mali_platform_device, mali_osk_resource_bank[i].irq_name);
```
用platform_get_irq_byname吧，看上去只是会有报错日志，没有额外的副作用

RadxaYuntian reviewed (changes requested):
提交整理一下吧

CodeChenL committed:
mali400: mali: fix compilation errors on kernel 6.1

```
ERROR: modpost: module mali uses symbol dma_buf_attach from namespace DMA_BUF, but does not import it.
ERROR: modpost: module mali uses symbol dma_buf_map_attachment from namespace DMA_BUF, but does not import it.
ERROR: modpost: module mali uses symbol dma_buf_unmap_attachment from namespace DMA_BUF, but does not import it.
ERROR: modpost: module mali uses symbol dma_buf_get from namespace DMA_BUF, but does not import it.
ERROR: modpost: module mali uses symbol dma_buf_put from namespace DMA_BUF, but does not import it.
ERROR: modpost: module mali uses symbol dma_buf_detach from namespace DMA_BUF, but does not import it.
```

Signed-off-by: ChenJiali <chenjiali@radxa.com>

RadxaYuntian commented on arch/arm64/boot/dts/rockchip/rk3528.dtsi:
```diff
@@ -1012,6 +1012,128 @@
+		operating-points-v2 = <&gpu_opp_table>;
+		#cooling-cells = <2>;
+		rockchip,grf = <&grf>;
+		status = "disabled";
```
这个是在板级设备树里面开启是吧？

CodeChenL replied:
是的

RadxaStephen commented:
需要处理一下编译报错

```
  CC      drivers/block/zram/zram_drv.o
drivers/gpu/arm/mali400/mali/linux/mali_osk_mali.c: In function '_mali_osk_resource_initialize':
drivers/gpu/arm/mali400/mali/linux/mali_osk_mali.c:126:19: error: unused variable 'res' [-Werror=unused-variable]
  126 |  struct resource *res;
      |                   ^~~
  CC      drivers/misc/lkdtm/bugs.o
cc1: all warnings being treated as errors
make[6]: *** [scripts/Makefile.build:250: drivers/gpu/arm/mali400/mali/linux/mali_osk_mali.o] Error 1
make[5]: *** [scripts/Makefile.build:500: drivers/gpu/arm/mali400/mali] Error 2
make[4]: *** [scripts/Makefile.build:500: drivers/gpu/arm/mali400] Error 2
make[4]: *** Waiting for unfinished jobs....
  CC      drivers/mfd/rkx110_x120/hal/cru_rkx110.o
```

CodeChenL committed:
mali400: mali: using platform_get_irq_byname to get interrupts

Due to API changes:
https://lore.kernel.org/lkml/20220316200633.28974-1-prabhakar.mahadev-lad.rj@bp.renesas.com/T/,
platform_get_resource_byname is no longer available for interrupts

Signed-off-by: ChenJiali <chenjiali@radxa.com>
